edge/pkg/devicetwin/dtmanager/communicate_test.go

```diff
@@ -0,0 +1,336 @@
+			context:  dtContext,
+			resource: "",
+			msg:      "",
+			wantErr:  errors.New("msg not Message type"),
+		},
+		{
+			name:     "dealLifeCycleTest-CloudConnected",
+			context:  dtContext,
+			resource: "",
+			msg:      &model.Message{Content: model.CloudConnected},
+			wantErr:  nil,
+		},
+		{
+			name:     "dealLifeCycleTest-CloudNotConnected",
+			context:  dtContext,
+			resource: "",
+			msg:      &model.Message{Content: model.CloudDisconnected},
+			wantErr:  nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := dealLifeCycle(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealLifeCycle() error = %v, wantErr %v", err, test.wantErr)
+			}
+		})
+	}
+}
+
+// TestDealConfirm is function to test dealConfirm().
+func TestDealConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		wantErr  error
+	}{
+		{
+			name:    "dealConfirmTest-WrongMsg",
+			context: dtContext,
+			msg:     "",
+			wantErr: errors.New("CommModule deal confirm, type not correct"),
+		},
+		{
+			name:    "dealConfirmTest-CorrectMsg",
+			context: dtContext,
+			msg:     &model.Message{Header: model.MessageHeader{ID: "id", ParentID: "parentId"}},
+			wantErr: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := dealConfirm(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealConfirm() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if err == nil {
+				_, exist := test.context.ConfirmMap.Load("parentId")
+				if exist {
+					t.Errorf("dealConfirm failed() ParentMessageId still present in context ConfirmMap")
+				}
+			}
+		})
+	}
+}
+
+// TestCheckConfirm is function to test checkConfirm().
+func TestCheckConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	dtContext.State = dtcommon.Connected
+	dtContext.ConfirmMap.Store("emptyMessage", &dttype.DTMessage{})
+	dtContext.ConfirmMap.Store("actionMessage", &dttype.DTMessage{Msg: &model.Message{}, Action: dtcommon.SendToCloud})
+	tests := []struct {
+		name    string
+		Worker  Worker
+		Group   string
+		context *dtcontext.DTContext
+		msg     interface{}
+	}{
+		{
+			name:    "checkConfirmTest",
+			Worker:  Worker{DTContexts: dtContext},
+			Group:   "",
+			context: dtContext,
+			msg:     &dttype.DTMessage{Action: "action"},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cw := CommWorker{
+				Worker: test.Worker,
+				Group:  test.Group,
+			}
+			cw.checkConfirm(test.context, test.msg)
+			for _, test := range tests {
```

Comment:
Do we need to loop twice over the tests array ? we also have a nested t.Run.....

Reply:
Updated